test/test_multidevice_ir.cpp

```diff
@@ -0,0 +1,136 @@
+#include <fusion.h>
+#include <fusion_segmenter.h>
+#include <gtest/gtest.h>
+#include <multidevice/executor.h>
+#include <multidevice/utils.h>
+#include <ops/all_ops.h>
+#include <test/multidevice.h>
+#include <test/validator.h>
+
+namespace nvfuser {
+TEST_F(MultiDeviceTest, ShardOuterAxisConcrete) {
+  int sharded_dim = 0;
+  std::unique_ptr<Fusion> fusion = std::make_unique<Fusion>();
+  FusionGuard fg(fusion.get());
+  int num_devices = communicator->size();
+  std::vector<int64_t> ranks(num_devices);
+  std::iota(ranks.begin(), ranks.end(), 0);
+  DeviceMesh mesh(ranks);
+
+  TensorView* tv0 = makeConcreteTensor({num_devices, 3});
+  TensorView* tv1 = add(tv0, tv0);
+  TensorView* tv2 = set(tv1);
+  TensorView* tv3 = add(tv2, tv2);
+  TensorView* tv4 = set(tv3);
+  TensorView* tv5 = sum(tv4, {0});
+  fusion->addInput(tv0);
+  fusion->addOutput(tv5);
+
+  // TODO: split
+  // tv3->split(sharded_dim, num_devices, false);
+  tv2->axis(sharded_dim)->parallelize(ParallelType::DIDx);
+  // tv3->split(sharded_dim, num_devices, false);
+  tv3->axis(sharded_dim)->parallelize(ParallelType::DIDx);
+
+  std::vector<TensorView*> tvs = {tv0, tv1, tv2, tv3, tv4, tv5};
+  for (auto tv : tvs) {
+    tv->setDeviceMesh(mesh);
+  }
+
+  auto x = at::randn({num_devices, 3}, tensor_options);
+  std::vector<c10::IValue> inputs = {x};
+  auto ref_outputs = at::sum(x * 4, {0});
+
+  MultiDeviceExecutor runtime(std::move(fusion), *communicator);
+  auto outputs = runtime.runWithInput(inputs);
+  testValidate(
+      runtime.fusion(), outputs, inputs, {ref_outputs}, __LINE__, __FILE__);
+}
+TEST_F(MultiDeviceTest, ShardOuterAxis) {
+  int sharded_dim = 0;
+  std::unique_ptr<Fusion> fusion = std::make_unique<Fusion>();
+  FusionGuard fg(fusion.get());
+  int num_devices = communicator->size();
+  std::vector<int64_t> ranks(num_devices);
+  std::iota(ranks.begin(), ranks.end(), 0);
+  DeviceMesh mesh(ranks);
+
+  TensorView* tv0 = makeContigTensor(2);
+  TensorView* tv1 = add(tv0, tv0);
+  TensorView* tv2 = set(tv1);
+  TensorView* tv3 = add(tv2, tv2);
+  TensorView* tv4 = set(tv3);
+  TensorView* tv5 = sum(tv4, {0});
+  fusion->addInput(tv0);
+  fusion->addOutput(tv5);
+
+  // TODO: split
+  // tv3->split(sharded_dim, num_devices, false);
+  tv2->axis(sharded_dim)->parallelize(ParallelType::DIDx);
+  // tv3->split(sharded_dim, num_devices, false);
+  tv3->axis(sharded_dim)->parallelize(ParallelType::DIDx);
+
+  std::vector<TensorView*> tvs = {tv0, tv1, tv2, tv3, tv4, tv5};
+  for (auto tv : tvs) {
+    tv->setDeviceMesh(mesh);
+  }
+
+  auto x = at::randn({num_devices, 3}, tensor_options);
+  std::vector<c10::IValue> inputs = {x};
+  auto ref_outputs = at::sum(x * 4, {0});
+
+  MultiDeviceExecutor runtime(std::move(fusion), *communicator);
+  auto outputs = runtime.runWithInput(inputs);
+  testValidate(
+      runtime.fusion(), outputs, inputs, {ref_outputs}, __LINE__, __FILE__);
+}
+
+inline at::Tensor shardInputTensor(at::Tensor tensor, int deviceId) {
```

Comment:
I think there is a small issue here. the index for slicing shouldn't be deviceId but the index of deviceId in the device mesh associated with the input. Here it works because, by chance, with the chosen device mesh, those two indices are equal.

But consider the device mesh
dm = {2,1}
then, device 1 should get the slice 0
device 2 should get the slice 1
and device 0 and others should get ... ? probably an empty vector. But from what I understand the system will complain if a has an empty vector.

Maybe this should be fixed, to make sure that tensors on devices where it is not sharded are empty tensors

What do you think ?

Reply:
thats a detail but in term of nvFuser's codestyle, every static declaration (i.e. that is not linked with an external file) should be placed in an anonymous namespace